frontend/lib/src/components/widgets/DateInput/DateInput.tsx

```diff
@@ -74,361 +59,312 @@ function datesToStrings(dates: Date[]): string[] {
-      element,
-      datesToStrings(this.state.values),
-      source,
-      fragmentId
-    )
-  }
-
+function DateInput({
+  disabled,
+  element,
+  widgetMgr,
+  width,
+  fragmentId,
+}: Props): ReactElement {
   /**
-   * If we're part of a clear_on_submit form, this will be called when our
-   * form is submitted. Restore our default value and update the WidgetManager.
+   * An array with start and end date specified by the user via the UI. If the user
+   * didn't touch this widget's UI, the default value is used. End date is optional.
    */
-  private onFormCleared = (): void => {
-    const defaultValue = stringsToDates(this.props.element.default)
-    this.setState(
-      {
-        values: defaultValue,
-        isEmpty: !defaultValue,
-      },
-      () => this.commitWidgetValue({ fromUi: true })
-    )
-  }
-
-  private handleChange = ({
-    date,
-  }: {
-    date: Date | (Date | null | undefined)[] | null | undefined
-  }): void => {
-    if (date === null || date === undefined) {
-      this.setState({
-        values: [],
-        isEmpty: true,
-      })
-      return
-    }
-
-    const values: Date[] = []
-    if (Array.isArray(date)) {
-      date.forEach((dt: Date | null | undefined) => {
-        if (dt) {
-          values.push(dt)
-        }
-      })
-    } else {
-      values.push(date)
-    }
-
-    this.setState(
-      {
-        values,
-        isEmpty: !values,
-      },
-      () => {
-        if (!this.state.isEmpty) this.commitWidgetValue({ fromUi: true })
+  const [value, setValueWithSource] = useBasicWidgetState<
+    Date[],
+    DateInputProto
+  >({
+    getStateFromWidgetMgr,
+    getDefaultStateFromProto,
+    getCurrStateFromProto,
+    updateWidgetMgrState,
+    element,
+    widgetMgr,
+    fragmentId,
+  })
+
+  const [isEmpty, setIsEmpty] = useState(false)
+
+  const theme = useTheme()
+  const { colors, fontSizes, lineHeights, spacing } = theme
+
+  const style = { width }
+  const minDate = moment(element.min, DATE_FORMAT).toDate()
+  const maxDate = getMaxDate(element)
+  const clearable = element.default.length === 0 && !disabled
+
+  // We need to extract the mask and format (date-fns notation) from the provided format string
+  // The user configured date format is based on the momentJS notation and is only allowed to contain
+  // one of YYYY/MM/DD, DD/MM/YYYY, or MM/DD/YYYY" and can also use a period (.) or hyphen (-) as separators.
+
+  // We need to convert the provided format into a mask supported by the Baseweb datepicker
+  // Thereby, we need to replace all letters with 9s which refers to any number.
+  const dateMask = element.format.replaceAll(/[a-zA-Z]/g, "9")
+  // The Baseweb datepicker supports the date-fns notation for date formatting which is
+  // slightly different from the momentJS notation. Therefore, we need to
+  // convert the provided format into the date-fns notation:
+  const dateFormat = element.format.replaceAll("Y", "y").replaceAll("D", "d")
```

Comment:
nit: I could imagine that `replaceAll` with a regex is on the more expensive side of operations; should we wrap it in a `useMemo` call or so?